### DISCUSSION TITLE
Refactor the APIs

### DISCUSSION THREAD

**Discussion by iskey** [eclipse-xpanse/xpanse#736] [General]
As-Is:

```
# 1. APIs

## 1.1. Monitor

- GET	 /xpanse/monitor/metric/service/{id}
- GET	 /xpanse/monitor/metric/resource/{id}

## 1.2. Service Vendor

- GET	 /xpanse/services/register/{id}
- GET	 /xpanse/services/register
- DELETE /xpanse/services/register/{id}
- POST	 /xpanse/services/register
- PUT	 /xpanse/services/register/{id}
- POST	 /xpanse/services/register/file
- PUT	 /xpanse/services/register/file/{id}

## 1.3. Service （APIs to manage the service instances）

- POST	 /xpanse/services/deploy
- GET	 /xpanse/services/deployed
- GET	 /xpanse/services/deployed/{id}/details/{userName}
- GET	 /xpanse/services/deployed/{userName}
- DELETE /xpanse/services/destroy/{id}- 

## Auth Management 

- GET /auth/token
- GET /auth/authorize

## Service Catalog (APIs to query the services which are available for the user to order)

- GET	 /xpanse/services/available
- GET	 /xpanse/services/available/{id}
- GET	 /xpanse/services/available/{id}/openapi
- GET	 /xpanse/services/available/category/{categoryName}- 


## Credentials Management (APIs for managing user's cloud provider credentials)

- PUT	 /xpanse/auth/csp/credential
- POST	 /xpanse/auth/csp/credential
- GET	 /xpanse/auth/user/credentials
- GET	 /xpanse/auth/csp/{cspName}/credentials
- DELETE /xpanse/auth/csp/{cspName}/credential
- GET	 /xpanse/auth/csp/{cspName}/credential/types
 GET	 /xpanse/auth/csp/{cspName}/credential/capabilities
- GET	 /xpanse/auth/csp/{cspName}/openapi/{type}

## Admin (APIs for administrating Xpanse)

- GET	 /xpanse/health

```

To-Be：

```
# 1. APIs

## 1.1. Monitor

- GET 	/xpanse/metrics?service_id="xxx"&resource_id="yyy"

## 1.2. Service Vendor

- GET	 /xpanse/service_templates/{id}
- GET	 /xpanse/service_templates
- DELETE  /xpanse/service_templates/{id}
- POST	 /xpanse/service_templates
- PUT	 /xpanse/service_templates/{id}
- POST	 /xpanse/service_templates/file
- PUT	 /xpanse/service_templates/file/{id}

## 1.3. Service （APIs to manage the service instances）

- POST	 /xpanse/services
- GET	 /xpanse/services
- GET	 /xpanse/service/{id}
- DELETE /xpanse/service/{id} 

## Auth Management 

- GET /auth/token
- GET /auth/authorize

## Service Catalog (APIs to query the services which are available for the user to order)

- GET	 /xpanse/catalog/services?categoryName="xxx"
- GET	 /xpanse/catalog/services/{id}
- GET	 /xpanse/catalog/services/{id}/openapi


## Credentials Management (APIs for managing user's cloud provider credentials)

- PUT	 /xpanse/credentials
- POST	 /xpanse/credentials
- GET	 /xpanse/credentials?csp="xxx"
- DELETE   /xpanse/credentials?csp="xxxx"&&type="xxxx"&&name="xxxx"

- GET	 /xpanse/credentials/capabilities?csp="huawei"

- GET	 /xpanse/credentials/openapi/{csp}/{type}

- GET	 /xpanse/credential_types?csp="xxx"

## Admin (APIs for administrating Xpanse)

- GET	 /xpanse/health

```

**Comment by baixinsui** [comment]
- PUT	 /xpanse/credential/{id}
- DELETE /xpanse/credential/{id}
The credential store in cache with key joined by csp,type,name,userId, there is no id for it.

**Comment by iskey** [comment]
> * PUT	 /xpanse/credential/{id}
> * DELETE /xpanse/credential/{id}
>   The credential store in cache with key joined by csp,type,name,userId, there is no id for it.

@swaroopar  @baixinsui  How about we remove the id, and put the `csp,type,name...` to the request body?

**Comment by swaroopar** [comment]
> > * PUT	 /xpanse/credential/{id}
> > * DELETE /xpanse/credential/{id}
> >   The credential store in cache with key joined by csp,type,name,userId, there is no id for it.
> 
> @swaroopar @baixinsui How about we remove the id, and put the `csp,type,name...` to the request body?

@iskey I think DELETE with a request body is not very normal in REST. Some discussion you can find here - https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request

**Comment by iskey** [comment]
> > > * PUT	 /xpanse/credential/{id}
> > > * DELETE /xpanse/credential/{id}
> > >   The credential store in cache with key joined by csp,type,name,userId, there is no id for it.
> > 
> > 
> > @swaroopar @baixinsui How about we remove the id, and put the `csp,type,name...` to the request body?
> 
> @iskey I think DELETE with a request body is not very normal in REST. Some discussion you can find here - https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request

Thank you for this @swaroopar , DELETE with a request body seems not suitable, so do you think adding the field `id` for the credential is a good idea?

**Comment by swaroopar** [comment]
@iskey the user ID itself is not needed. Wee have this value already in the header.

**Comment by baixinsui** [comment]
## 1.2. Service Vendor
- GET /xpanse/services/register/{id}    -> GET /xpanse/service_template/{id}

@iskey @swaroopar  Except using 'service_template' to replace 'register' in the uri , do we still need to make the following modifications: module 'register' ->module 'templates'? DB entity 'RegisterServiceEntity' ->'ServiceTemplateEntity'?  Object in models like 'RegisterServiceVo' -> 'ServiceTemplateVO' ?

**Comment by swaroopar** [comment]
@baixinsui yes I would say so.. even the module must be service template instead of just template.

**Comment by baixinsui** [comment]
Ok.

**Comment by baixinsui** [comment]
I will do it.

**Comment by baixinsui** [comment]
I have completed 'Refactor the APIs' in the backend-API. Due to too many code had been modified, I will split multiple issues  according to the ApiController to complete the entire task and complete UI-client and backend-API modification and testing in each smaller issue for each ApiController.

**Comment by iskey** [comment]
As we have created some sub-issues to complete this refactor, I will convert this issue to a discussion for an archive.
